### ISSUE TITLE
Errors when running garbage collection

### ISSUE THREAD

**Issue by treydock** [goharbor/harbor#15039]
If you are reporting a problem, please make sure the following information are provided:

**Expected behavior and actual behavior:**
I am getting errors during both manual and scheduled garbage collection:

```
2021-06-01T13:00:16Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:271]: delete manifest from storage: sha256:ccf1e1ce8307193814b206c6cf2cb5017dddc69b333df9e47f5e5a63edc14db8
2021-06-01T13:00:16Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:283]: delete artifact trash record from database: 8, gateways/phylogatr, sha256:ccf1e1ce8307193814b206c6cf2cb5017dddc69b333df9e47f5e5a63edc14db8
2021-06-01T13:00:16Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:295]: delete blob from storage: sha256:ccf1e1ce8307193814b206c6cf2cb5017dddc69b333df9e47f5e5a63edc14db8
2021-06-01T13:00:16Z [ERROR] [/jobservice/job/impl/gc/garbage_collection.go:165]: failed to execute GC job at sweep phase, error: failed to delete blob from storage: sha256:ccf1e1ce8307193814b206c6cf2cb5017dddc69b333df9e47f5e5a63edc14db8, deletefailed: http status code: 500, body: {"errors":[{"code":"UNKNOWN","message":"internal server error"}]}
```

This is log from `registryct.log`

```
Jun  1 09:00:16 172.23.0.1 registryctl[1693]: time="2021-06-01T13:00:16Z" level=info msg="Deleting blob: /docker/registry/v2/blobs/sha256/cc/ccf1e1ce8307193814b206c6cf2cb5017dddc69b333df9e47f5e5a63edc14db8" go.version=go1.15.6
Jun  1 09:00:16 172.23.0.1 registryctl[1693]: 2021-06-01T13:00:16Z [INFO] [/registryctl/api/registry/blob/blob.go:44]: failed to remove blob: sha256:ccf1e1ce8307193814b206c6cf2cb5017dddc69b333df9e47f5e5a63edc14db8, with error:filesystem: unlinkat /storage/docker/registry/v2/blobs/sha256/cc/ccf1e1ce8307193814b206c6cf2cb5017dddc69b333df9e47f5e5a63edc14db8: directory not empty
Jun  1 09:00:16 172.23.0.1 registryctl[1693]: 2021-06-01T13:00:16Z [ERROR] [/lib/http/error.go:54]: {"errors":[{"code":"UNKNOWN","message":"unknown: filesystem: unlinkat /storage/docker/registry/v2/blobs/sha256/cc/ccf1e1ce8307193814b206c6cf2cb5017dddc69b333df9e47f5e5a63edc14db8: directory not empty"}]} 
Jun  1 09:00:16 172.23.0.1 registryctl[1693]: 172.23.0.11 - - [01/Jun/2021:13:00:16 +0000] "DELETE /api/registry/blob/sha256:ccf1e1ce8307193814b206c6cf2cb5017dddc69b333df9e47f5e5a63edc14db8 HTTP/1.1" 500 66
```

**Steps to reproduce the problem:**
Please provide the steps to reproduce this problem.

Run `GC Now` and wait for error to pop up after a little bit of time.

**Versions:**
Please specify the versions of following systems.

- harbor version: v2.2.0-ec0ba116
- docker engine version: 20.10.3
- docker-compose version: 1.25.1

**Additional context:**

This is Harbor running everything within docker compose deployment, nothing external, and on RHEL 7.9 with the data directory on NFSv3.

This is my configuration:

```yaml
# Configuration file of Harbor

# The IP address or hostname to access admin UI and registry service.
# DO NOT use localhost or 127.0.0.1, because Harbor needs to be accessed by external clients.
hostname: docker-registry.OMIT

# http related config
http:
  # port for http, default is 80. If https enabled, this port will redirect to https port
  port: 80

# https related config
https:
  # https port for harbor, default is 443
  port: 443
  # The path of cert and key files for nginx
  certificate: /etc/pki/tls/certs/docker-registry.OMIT-full.crt
  private_key: /etc/pki/tls/private/docker-registry.OMIT.key


# # Uncomment following will enable tls communication between all harbor components
internal_tls:
#   # set enabled to true means internal tls is enabled
  enabled: true
#   # put your cert and key files on dir
  dir: /data/internal-tls


# The initial password of Harbor admin
# It only works in first time to install harbor
# Remember Change the admin password from UI after launching Harbor.
harbor_admin_password: OMIT

# Harbor DB configuration
database:
  # The password for the root user of Harbor DB. Change this before any production use.
  password: OMIT
  # The maximum number of connections in the idle connection pool. If it <=0, no idle connections are retained.
  max_idle_conns: 50
  # The maximum number of open connections to the database. If it <= 0, then there is no limit on the number of open connections.
  # Note: the default number of connections is 100 for postgres.
  max_open_conns: 100

# The default data volume
data_volume: /data

# Harbor Storage settings by default is using /data dir on local filesystem
# Uncomment storage_service setting If you want to using external storage
# storage_service:
#   # ca_bundle is the path to the custom root ca certificate, which will be injected into the truststore
#   # of registry's and chart repository's containers.  This is usually needed when the user hosts a internal storage with self signed certificate.
#   ca_bundle:

#   # storage backend, default is filesystem, options include filesystem, azure, gcs, s3, swift and oss
#   # for more info about this configuration please refer https://docs.docker.com/registry/configuration/
#   filesystem:
#     maxthreads: 100
#   # set disable to true when you want to disable registry redirect
#   redirect:
#     disabled: false

# Trivy configuration
#
# Trivy DB contains vulnerability information from NVD, Red Hat, and many other upstream vulnerability databases.
# It is downloaded by Trivy from the GitHub release page https://github.com/aquasecurity/trivy-db/releases and cached
# in the local file system. In addition, the database contains the update timestamp so Trivy can detect whether it
# should download a newer version from the Internet or use the cached one. Currently, the database is updated every
# 12 hours and published as a new release to GitHub.
trivy:
  # ignoreUnfixed The flag to display only fixed vulnerabilities
  ignore_unfixed: false
  # skipUpdate The flag to enable or disable Trivy DB downloads from GitHub
  #
  # You might want to enable this flag in test or CI/CD environments to avoid GitHub rate limiting issues.
  # If the flag is enabled you have to download the `trivy-offline.tar.gz` archive manually, extract `trivy.db` and
  # `metadata.json` files and mount them in the `/home/scanner/.cache/trivy/db` path.
  skip_update: false
  #
  # insecure The flag to skip verifying registry certificate
  insecure: false
  # github_token The GitHub access token to download Trivy DB
  #
  # Anonymous downloads from GitHub are subject to the limit of 60 requests per hour. Normally such rate limit is enough
  # for production operations. If, for any reason, it's not enough, you could increase the rate limit to 5000
  # requests per hour by specifying the GitHub access token. For more details on GitHub rate limiting please consult
  # https://developer.github.com/v3/#rate-limiting
  #
  # You can create a GitHub token by following the instructions in
  # https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line
  #

# Clair configuration
clair: 
  # The interval of clair updaters, the unit is hour, set to 0 to disable the updaters.
  updaters_interval: 12

jobservice:
  # Maximum number of job workers in job service  
  max_job_workers: 10

notification:
  webhook_job_max_retry: 10

chart:
  # Change the value of absolute_url to enabled can enable absolute url in chart
  absolute_url: absolute_url

# Log configurations
log:
  # options are debug, info, warning, error, fatal
  level: info
  # configs for logs in local storage
  local:
    # Log files are rotated log_rotate_count times before being removed. If count is 0, old versions are removed rather than rotated.
    rotate_count: 50
    # Log files are rotated only if they grow bigger than log_rotate_size bytes. If size is followed by k, the size is assumed to be in kilobytes. 
    # If the M is used, the size is in megabytes, and if G is used, the size is in gigabytes. So size 100, size 100k, size 100M and size 100G 
    # are all valid.
    rotate_size: 200M
    # The directory on your host that store log
    location: /var/log/harbor

  # Uncomment following lines to enable external syslog endpoint.

#This attribute is for migrator to detect the version of the .cfg file, DO NOT MODIFY!
_version: 2.2.0

# Uncomment external_database if using external database. Currently only support POSTGRES. 
# Four databases are needed to be create first by users for Harbor core, Clair, Notary server 
# and Notary signer. And the tables will be generated automatically when Harbor starting up.
# NOTE: external_database is unable to custom attributes individually, you must do them in block.
# external_database:
#   harbor:
#     host: harbor_db_host
#     port: harbor_db_port
#     db_name: harbor_db_name
#     username: harbor_db_username
#     password: harbor_db_password
#     ssl_mode: disable
#   clair:
#     host: clair_db_host
#     port: clair_db_port
#     db_name: clair_db_name
#     username: clair_db_username
#     password: clair_db_password
#     ssl_mode: disable
#   notary_signer:
#     host: notary_signer_db_host
#     port: notary_signer_db_port
#     db_name: notary_signer_db_name
#     username: notary_signer_db_username
#     password: notary_signer_db_password
#     ssl_mode: disable
#   notary_server:
#     host: notary_server_db_host
#     port: notary_server_db_port
#     db_name: notary_server_db_name
#     username: notary_server_db_username
#     password: notary_server_db_password
#     ssl_mode: disable





# Global proxy
# Config http proxy for components, e.g. http://my.proxy.com:3128
# Components doesn't need to connect to each others via http proxy.
# Remove component from `components` array if want disable proxy
# for it. If you want use proxy for replication, MUST enable proxy
# for core and jobservice, and set `http_proxy` and `https_proxy`.
# Add domain to the `no_proxy` field, when you want disable proxy
# for some special registry.
proxy:
  http_proxy: 
  https_proxy: 
  # no_proxy endpoints will appended to 127.0.0.1,localhost,.local,.internal,log,db,redis,nginx,core,portal,postgresql,jobservice,registry,registryctl,clair,chartmuseum,notary-server
  no_proxy: 127.0.0.1,localhost,.local,.internal,log,db,redis,nginx,core,portal,postgresql,jobservice,registry,registryctl,clair
  components:
    - core
    - jobservice
    - clair
metric:
  enabled: true
  port: 9090
  path: /metrics
```

**Comment by treydock** [comment]
I manually inspected the filesystem path mentioned in logs and the directory was in fact empty, so I re-ran GC and the operation was successful.  I did see the error twice. I had the scheduled GC run weekly and this is all I had in logs:

```
{
"errors": [
{
"code": "UNKNOWN",
"message": "internal server error"
}
]
}
```

That ran on a Sunday, the following Tuesday I ran `GC Now` and got a similar error except this time the GC logs had actual log messages but also indicated `internal server error`. That's when I noticed the path in registryctl.log and inspected things and ran `GC Now` again and things were successful.
